src/Workspaces/Core/Portable/Workspace/Solution/DocumentInfo.cs

```diff
@@ -191,14 +188,14 @@ internal class DocumentAttributes : IChecksummedObject, IObjectWritable
             public DocumentAttributes(
                 DocumentId id,
                 string name,
-                IEnumerable<string>? folders,
+                IReadOnlyList<string> folders,
                 SourceCodeKind sourceCodeKind,
                 string? filePath,
                 bool isGenerated)
             {
-                Id = id ?? throw new ArgumentNullException(nameof(id));
-                Name = name ?? throw new ArgumentNullException(nameof(name));
-                Folders = folders.ToImmutableReadOnlyListOrEmpty();
```

Comment:
is this not a better place for the throws? i worry that someone could now call this and pass in null and not have it be caught.

Reply:
This is not public APIs.
Internally we use nullable annotations, so that should guard from internal code calling with null. 
Public APIs should validate their inputs directly. 

I can add `Debug.Assert` here.

Reply:
I'd be fine with a Contract call here.  We'd def want ot catch this in release as well as debug.  if we have that, i'm ok with the change.